test/utils/submission_queue.test.ts

```diff
@@ -50,16 +61,82 @@ describe('SubmissionQueue', () => {
     const retrySubmission = jest.spyOn(queue as any, 'retrySubmission' as any);
     const onDrop = jest.spyOn(queue as any, 'onDrop' as any);
     const submission = { name: 'submission', ack: false };
-    queue.addSubmission(submission, 3);
+    queue.addSubmission(submission, 3, 100);
+    while (queue.processing) {
+      await new Promise((resolve) => setTimeout(resolve, 50));
+    }
+    expect(queue.submissions.length).toBe(0);
+    // 3 times to retry + 1 time to drop
+    expect(retrySubmission).toHaveBeenCalledTimes(4);
+    expect(logger.error).toHaveBeenCalledWith(
+      'Submission retry limit reached, dropping submission: ' + JSON.stringify(submission)
+    );
+    expect(onDrop).toHaveBeenCalledWith(submission);
+  });
+
+  test('should retry entries when not acknowledged and drop entries after max retries', async () => {
+    const retrySubmission = jest.spyOn(queue as any, 'retrySubmission' as any);
+    const onDrop = jest.spyOn(queue as any, 'onDrop' as any);
+    const submission = { name: 'submission', ack: false };
+    let timer_start = Date.now();
+    queue.addSubmission(submission, 3, 100);
     while (queue.processing) {
       await new Promise((resolve) => setTimeout(resolve, 50));
     }
+    let time_passed = Date.now() - timer_start;
     expect(queue.submissions.length).toBe(0);
     // 3 times to retry + 1 time to drop
     expect(retrySubmission).toHaveBeenCalledTimes(4);
     expect(logger.error).toHaveBeenCalledWith(
       'Submission retry limit reached, dropping submission: ' + JSON.stringify(submission)
     );
     expect(onDrop).toHaveBeenCalledWith(submission);
+    expect(time_passed).toBeGreaterThanOrEqual(300);
+    expect(time_passed).toBeLessThanOrEqual(400);
+  });
+
+  test('should respect minRetryTimeout', async () => {
+    const retrySubmission = jest.spyOn(queue as any, 'retrySubmission' as any);
+    const onDrop = jest.spyOn(queue as any, 'onDrop' as any);
+    const submission = { name: 'submission', ack: false };
+    let timer_start = Date.now();
+    queue.addSubmission(submission, 3, 500);
+    while (queue.processing) {
+      await new Promise((resolve) => setTimeout(resolve, 50));
+    }
+    let time_passed = Date.now() - timer_start;
+    expect(queue.submissions.length).toBe(0);
+    // 3 times to retry + 1 time to drop
+    expect(retrySubmission).toHaveBeenCalledTimes(4);
+    expect(logger.error).toHaveBeenCalledWith(
+      'Submission retry limit reached, dropping submission: ' + JSON.stringify(submission)
+    );
+    expect(onDrop).toHaveBeenCalledWith(submission);
+    expect(time_passed).toBeGreaterThanOrEqual(1500);
+    expect(time_passed).toBeLessThanOrEqual(1600);
+  });
```

Comment:
Is this test the same as the previous?

Reply:
no, the time changed, since I'm validating it with actual system time, using two tests was the easiest way to verify it worked as expected